exchange/exchange.go

```diff
@@ -0,0 +1,221 @@
+	e.adapters = make([]openrtb_ext.BidderName, 0, len(e.adapterMap))
+	for a, _ := range e.adapterMap {
+		e.adapters = append(e.adapters, a)
+	}
+	return e
+}
+
+func (e *exchange) HoldAuction(ctx context.Context, bidRequest *openrtb.BidRequest) (*openrtb.BidResponse, error) {
+	// Slice of BidRequests, each a copy of the original cleaned to only contain bidder data for the named bidder
+	cleanRequests, errs := openrtb_ext.CleanOpenRTBRequests(bidRequest, e.adapters)
+	// List of bidders we have requests for.
+	liveAdapters := make([]openrtb_ext.BidderName, len(cleanRequests))
+	i := 0
+	for a, _ := range cleanRequests {
+		liveAdapters[i] = a
+		i++
+	}
+	// Randomize the list of adapters to make the auction more fair
+	openrtb_ext.RandomizeList(liveAdapters)
+
+	adapterBids, adapterExtra := e.GetAllBids(ctx, liveAdapters, cleanRequests)
+
+	// Build the response
+	return e.BuildBidResponse(liveAdapters, adapterBids, bidRequest, adapterExtra, errs)
+}
+
+// This piece sends all the requests to the bidder adapters and gathers the results.
+func (e *exchange) GetAllBids(ctx context.Context, liveAdapters []openrtb_ext.BidderName, cleanRequests map[openrtb_ext.BidderName]*openrtb.BidRequest) (map[openrtb_ext.BidderName]*pbsOrtbSeatBid, map[openrtb_ext.BidderName]*seatResponseExtra) {
+	// Set up pointers to the bid results
+	adapterBids := map[openrtb_ext.BidderName]*pbsOrtbSeatBid{}
+	adapterExtra := make(map[openrtb_ext.BidderName]*seatResponseExtra)
+	chBids := make(chan *bidResponseWrapper, len(liveAdapters))
+	for _, a := range liveAdapters {
+		// Here we actually call the adapters and collect the bids.
+		go func(aName openrtb_ext.BidderName) {
+			// Passing in aName so a doesn't change out from under the go routine
+			brw := new(bidResponseWrapper)
+			brw.bidder = aName
+			start := time.Now()
+			bids, err := e.adapterMap[aName].requestBid(ctx, cleanRequests[aName])
+
+			// Add in time reporting
+			elapsed := time.Since(start)
+			brw.adapterBids = bids
+			// Structure to record extra tracking data generated during bidding
+			ae := new(seatResponseExtra)
+			ae.ResponseTimeMillis = int(elapsed/time.Millisecond)
+			serr := make([]string, len(err))
+			for i :=0; i<len(err); i++ {
+				serr[i] = err[i].Error()
+			}
+			ae.Errors = serr
+			brw.adapterExtra = ae
+			chBids <- brw
+		}(a)
+	}
+	// Wait for the bidders to do their thing
+	for i := 0; i < len(liveAdapters); i++ {
+		brw := <- chBids
+		adapterExtra[brw.bidder] = brw.adapterExtra
+		adapterBids[brw.bidder] = brw.adapterBids
+	}
+
+	return adapterBids, adapterExtra
+}
+
+// This piece takes all the bids supplied by the adapters and crafts an openRTB response to send back to the requester
+func (e *exchange) BuildBidResponse(liveAdapters []openrtb_ext.BidderName, adapterBids map[openrtb_ext.BidderName]*pbsOrtbSeatBid, bidRequest *openrtb.BidRequest, adapterExtra map[openrtb_ext.BidderName]*seatResponseExtra, errList []error) (*openrtb.BidResponse, error) {
```

Comment:
This is an interesting idea. My initial avoidance of a disposable object was the overhead of building up and tearing down an object for each auction. But then again if I am building up and tearing down all the elements of the struct anyway there should not be a real cost with doing it that way. I should review the code and verify this would not make more objects. The bid request should be a pointer, I never modify the incoming request, so there is no need to copy data from it into a new object.

Implementing this is a little messy at the moment. This will tie strongly into the targeting patch, so it may be best to create a branch from targeting to implement this change. That will mean a cascade of PRs to get openrtb into the final state.

If we want this change to go in ahead of targeting, I think the solution would be to implement it in the targeting branch, and then backport it into this branch. It will make the merges very messy, but keeps from having to rewrite this for targeting after going down a wrong path doing it just for the current branch.

Thoughts?

Reply:
Yeah.... this is why I hate building code off of open PRs. That's a familiar dilemma for me.

In this case, I wouldn't worry about it at all. We should keep this pattern in mind for the future, but I think your code here is fine the way it is.

If we had to, the least painful strategy is  actually to make the changes in _this_ branch, and then merge them back into your targeting branch and fix the conflicts there. If you manually make the same code change in two places (without `git` doing the merge), you'll get weird merge conflicts later.